feat(experimental): add worker session data model for job attachment sync operations

godobyte opened the pull request:
### What was the problem/requirement? (What/Why)
The worker agent needed a more structured way to manage job attachment data during session execution, specifically to track local file system paths and manifest information without relying solely on hash-based correlation. 

The existing system lacked a dedicated data model for worker-specific attachment operations, making it difficult to maintain state and coordinate between attachment upload/download operations and session management.

### What was the solution? (How)
- Added `WorkerManifestProperties` class: A new data structure that extends `ManifestProperties` with worker-specific local path information
- Enhanced Session class: Added methods to manage a dictionary of worker manifest properties keyed by local root paths

### What is the impact of this change?
- Provides a structured approach to managing worker-specific attachment data during session execution
- Enables better tracking of local file system paths and manifest locations
- Lays groundwork for improved job attachment sync data flow

### How was this change tested?
Added tests for the data model and session integration

### Was this change documented?
Yes added docstrings and inline comments.

### Is this a breaking change?
No

----

*By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.*

godobyte committed:
feat(experimental): add model for worker session and job attachment sync data flow

Signed-off-by: Godot Bian <13778003+godobyte@users.noreply.github.com>

mwiebe commented on src/deadline_worker_agent/sessions/attachment_models.py:
```diff
@@ -0,0 +1,175 @@
+                "input_manifest_path": self.manifest_properties.inputManifestPath,
+                "input_manifest_hash": self.manifest_properties.inputManifestHash,
+                "output_relative_directories": self.manifest_properties.outputRelativeDirectories,
+            },
```
The `ManifestProperties` class has a `to_dict` member (see https://github.com/aws-deadline/deadline-cloud/blob/67869c6ba09fce023fd37d0ee5e9981f1f4b83eb/src/deadline/job_attachments/models.py#L206). This could call that member to shorten the code and remove the duplication.

In `ManifestProperties.to_dict`, the dict/json form of these objects is using camel case instead of snake case, it would be nice to follow that convention in this `to_dict` as well.

```suggestion
            "manifestProperties": self.manifest_properties.to_dict(),
```

godobyte replied:
Thanks for the feedback. Let me explain my reasoning.

Regarding the naming convention, `ManifestProperties` currently uses camelCase because it's used for parsing API responses. Since there's no specific requirement for camelCase in this context, I opted for snake_case to maintain Python conventions.

I agree with your suggestion about using ManifestProperties.to_dict. The maintainability and consistency with the model outweigh the Python convention benefits. Future API changes would only need updates in one place, which reduces risk of de-synchronization between the model and its usage

I will update to use this approach.

mwiebe commented on src/deadline_worker_agent/sessions/session.py:
```diff
@@ -223,6 +224,7 @@ def __init__(
         self._executor = ThreadPoolExecutor(max_workers=1)
         self._manifest_paths_by_root = dict()
         self._manifest_out_rel_dirs_by_source: dict[str, list[str]] = dict()
+        self._worker_manifest_properties_dict: dict[str, WorkerManifestProperties] = {}
```
From this line, I can't tell what the key values of this dict represent. Below, I can see it using the local root path for the key. Following how the other properties here are named, `_worker_manifest_properties_by_local_root` would be a good name that communicates what the keys are and fits in well.

mwiebe commented on test/unit/sessions/test_session.py:
```diff
@@ -2698,3 +2701,326 @@ def test_session_includes_redacted_env_vars_extension(
+        assert result is not None
+        assert len(result.local_manifest_paths) == 2
+        assert "/first/manifest.json" in result.local_manifest_paths
+        assert "/second/manifest.json" in result.local_manifest_paths
```
Here's a way you can express these three lines, that gives helpful error messages in the logs when you get a test failure. By sorting the list, it normalizes it for comparison. By comparing the list to an expected value instead of checking the length, pytest will show you the difference between the two lists with a helpful diff syntax.

```suggestion
        assert sorted(result.local_manifest_paths) == ["/first/manifest.json", "/second/manifest.json"]
```

godobyte replied:
Thanks for the suggestion. I'll update the test to use this pattern, it'll definitely make troubleshooting easier when tests fail.

leongdl commented on src/deadline_worker_agent/sessions/attachment_models.py:
```diff
@@ -0,0 +1,175 @@
+            destination_path=self.local_root_path,
+        )
+
+    def get_hashed_source_path(self) -> str:
```
Will we refactor the hash data string formatting so this is clear who owns it and the set standard?

godobyte replied:
Yes, I agree this should be part of Job Attachment library. I've been looking for an appropriate place to define this contract, ideally within a data model.

The challenge lies in the mixed sources of the data: the hash algorithm comes from the manifest model, while 'file_system_location_name' and 'root_path' are job-level properties. For now, I've placed it here as a temporary solution.

leongdl replied:
ok - this should be in the comment then. Otherwise this code will get recycled and then impossible to remove to a proper interface.

godobyte replied:
Oh just saw the comment, I'll add it in a follow-up PR if I don't get to another rev on this PR.

mwiebe replied:
I suspect the key will be that a document will own this, not code. I think we'll want to put that in https://github.com/aws-deadline/deadline-cloud/tree/mainline/docs

godobyte replied:
Added a TODO below on L94 to track the update.

leongdl replied:
I suspect the key will be that a document will own this, not code. I think we'll want to put that in https://github.com/aws-deadline/deadline-cloud/tree/mainline/docs

IMO, adding to docs is great but not everyone will read or discover it or read every version of it. Only code and tests can discourage access.

leongdl commented on src/deadline_worker_agent/sessions/attachment_models.py:
```diff
@@ -0,0 +1,175 @@
+
+
+@dataclass
+class WorkerManifestProperties:
```
Nit - Would WorkerSessionActionManfiestProperties make sense, or WorkerSessionManifestProperty?

Since this data structure is closely tied to how a Worker's Session works.

godobyte replied:
I chose `WorkerManifestProperties` because it clearly describes its purpose: properties related to manifests in the worker context. Since we don't currently have other `Worker*ManifestProperties` classes, adding more specificity (like Session) might be redundant.

However, if the team feels adding Session-related concepts would make the naming more clear, I'm open to that change.

leongdl replied:
ok!

leongdl commented on src/deadline_worker_agent/sessions/attachment_models.py:
```diff
@@ -0,0 +1,175 @@
+        )
+
+    @classmethod
+    def from_manifest_properties(
```
Can you link where we will integrate this? I want to see how it will be used.

godobyte replied:
The collection of this is kept at session for state management, and written do disk for attachment_upload.py and attachment_download.py. `attachment_download` constructs the data, and `attachment_update` modifies and consumes it.

Here is also prototyping code if you'd like to read for more detail.
https://github.com/aws-deadline/deadline-cloud-worker-agent/compare/mainline...godobyte:deadline-cloud-worker-agent:wa-ja-prototype

jericht replied:
Do we need this function? Couldn't callers just instantiate this dataclass directly?

leongdl replied:
Where is here?

godobyte replied:
Sorry missed the linked, added to the comment above to indicate the usage. 

> Do we need this function? Couldn't callers just instantiate this dataclass directly?

Yes we'll parse from the API response directly, which contains ManifestProperties.

jusiskin replied:
+1 to what @jericht said. I think we can just use:

```py
WorkerManifestProperties(
  manifest_properties=...,
  local_manifest_paths=...,
  local_root_path=...
)
```

I'm not sure I see the value of having this `WorkerManifestProperties.from_manifest_properties(...)` class function. It is longer to type.

godobyte replied:
ah I might have misread the initial comment, I can update this.

moorec-aws committed:
chore: add responded and stale issue/pr workflows (#726)

Signed-off-by: Charles Moore <122481442+moorec-aws@users.noreply.github.com>
Signed-off-by: Godot Bian <13778003+godobyte@users.noreply.github.com>

jericht commented on src/deadline_worker_agent/sessions/attachment_models.py:
```diff
@@ -0,0 +1,175 @@
+        if not self.local_root_path:
+            raise ValueError("local_root_path must be set for WorkerManifestProperties")
+
+    @property
+    def root_path(self) -> str:
+        """Get the original root path from manifest properties."""
+        return self.manifest_properties.rootPath
+
+    @property
+    def root_path_format(self):
+        """Get the root path format from manifest properties."""
+        return self.manifest_properties.rootPathFormat
+
+    @property
+    def file_system_location_name(self) -> Optional[str]:
+        """Get the file system location name from manifest properties."""
+        return self.manifest_properties.fileSystemLocationName
+
+    @property
+    def input_manifest_path(self) -> Optional[str]:
+        """Get the input manifest path from manifest properties."""
+        return self.manifest_properties.inputManifestPath
+
+    @property
+    def input_manifest_hash(self) -> Optional[str]:
+        """Get the input manifest hash from manifest properties."""
+        return self.manifest_properties.inputManifestHash
+
+    @property
+    def output_relative_directories(self) -> Optional[List[str]]:
+        """Get the output relative directories from manifest properties."""
+        return self.manifest_properties.outputRelativeDirectories
```
Could we inherit from `ManifestProperties` instead of defining these properties for all of the underlying properties? Alternatively, we could just let callers reach into the underlying `manifest_properties` themselves if we want to keep this structure.

godobyte replied:
`ManifestProperties` class is defined with camelCase since it parses API result. The current approach is to follow python convention and avoid potential serialization issue. The property forwarding is to make it more clear and explicit.

jericht commented on src/deadline_worker_agent/sessions/session.py:
```diff
@@ -299,6 +301,74 @@ def add_manifest_out_rel_dirs(self, source: str, out_rel_dirs: list[str]):
+        """Worker-specific manifest properties dictionary for enhanced job attachment processing"""
+        return self._worker_manifest_properties_dict
+
+    def set_worker_manifest_properties(
+        self, worker_manifest_properties: WorkerManifestProperties
+    ) -> None:
+        """Add worker manifest properties to the cached dictionary
+
+        Parameters
+        ----------
+        worker_manifest_properties : WorkerManifestProperties
+            Worker manifest properties to add to the cached session dictionary
+        """
+        self._worker_manifest_properties_dict[worker_manifest_properties.local_root_path] = (
+            worker_manifest_properties
+        )
+
+    def get_worker_manifest_properties(
+        self, local_root_path: str
+    ) -> WorkerManifestProperties | None:
+        """Get worker manifest properties by local root path
+
+        Parameters
+        ----------
+        local_root_path : str
+            The local root path key to look up
+
+        Returns
+        -------
+        WorkerManifestProperties | None
+            The worker manifest properties for the given key, or None if not found
+        """
+        return self._worker_manifest_properties_dict.get(local_root_path)
```
Do we need these two functions? They are just wrapping around the `worker_manifest_properties_dict` object.

I can see the `set_worker_manifest_properties` being a little convenient because it pulls the `local_root_path` out of the manifest properties to key into the dict for us, but not sure it's worth a dedicated function.

godobyte replied:
Yes it'll be called in attachment download action, and it's not recommended to directly access a private member `self._worker_manifest_properties_dict`.

jericht commented on src/deadline_worker_agent/sessions/session.py:
```diff
@@ -299,6 +301,74 @@ def add_manifest_out_rel_dirs(self, source: str, out_rel_dirs: list[str]):
+        """
+        return self._worker_manifest_properties_dict.get(local_root_path)
+
+    def get_worker_manifest_properties_list(self) -> list[WorkerManifestProperties]:
+        """Get the cached worker manifest properties as a list
+
+        Returns
+        -------
+        list[WorkerManifestProperties]
+            The cached list of worker manifest properties
+        """
+        return list(self._worker_manifest_properties_dict.values())
```
Do we need this? Also just a wrapper around the `worker_manfiest_properties_dict`

godobyte replied:
Yes same idea as above for `get_worker_manifest_properties`.

jericht commented on src/deadline_worker_agent/sessions/session.py:
```diff
@@ -299,6 +301,74 @@ def add_manifest_out_rel_dirs(self, source: str, out_rel_dirs: list[str]):
+        """
+        return list(self._worker_manifest_properties_dict.values())
+
+    def add_local_manifest_path(self, local_root_path: str, manifest_path: str) -> None:
+        """Add a local manifest path to the worker manifest properties
+
+        Parameters
+        ----------
+        local_root_path : str
+            The local root path key to look up the worker manifest properties
+        manifest_path : str
+            The manifest path to add to the local_manifest_paths list
+
+        Raises
+        ------
+        ValueError
+            If the local_root_path is not found in the worker manifest properties dictionary
+        """
+        if worker_props := self._worker_manifest_properties_dict.get(local_root_path):
+            worker_props.local_manifest_paths.append(manifest_path)
+        else:
+            raise ValueError(
+                f"Worker manifest properties not found for local_root_path: {local_root_path}"
+            )
```
This also looks like a wrapper around the `_worker_manifest_properties_dict` as well. Do we need this?

e.g. a caller could just write this themselves:

```py
manifest_props = ... # a WorkerManifestProperties object
session = ... # a Session object
if manifest.local_root_path not in session.worker_manifest_properties_dict:
    # handle it.. e.g. skip the operation or raise an error
else:
    session.worker_manifest_properties_dict[manifest.local_root_path].append(manifest_path)
```

godobyte replied:
Same idea as above, we don't want to directly manipulate `session._worker_manifest_properties_dict` data via . operations and direct assignment.

godobyte committed:
fix linting error

Signed-off-by: Godot Bian <13778003+godobyte@users.noreply.github.com>

mwiebe commented on src/deadline_worker_agent/sessions/attachment_models.py:
```diff
@@ -0,0 +1,183 @@
+        if self.manifest_properties.outputRelativeDirectories:
+            manifest_dict["outputRelativeDirectories"] = (
+                self.manifest_properties.outputRelativeDirectories
+            )
```
The `to_dict` method is part of the public interface of the deadline-cloud library, it would be nice to use that directly.

See https://github.com/aws-deadline/deadline-cloud/blob/mainline/src/deadline/job_attachments/models.py#L206 to note that the `ManifestProperties` dataclass is available with the name `deadline.job_attachments.models.ManifestProperties`, and `to_dict` is a public method: there are no names prefixed with `_` in that chain.

See also https://github.com/aws-deadline/deadline-cloud/blob/mainline/DEVELOPMENT.md#public-contracts where we document:

> All functions and modules whose name does not begin with an underscore are part of the public contract for this package.

godobyte replied:
Ah you're right. That was my plan. I mixed that with `from_dict` for some reason during the implementation, which is not in the model. Updated to use `to_dict` from `ManifestProperties`.

godobyte committed:
fix to_dict to reuse from model

Signed-off-by: Godot Bian <13778003+godobyte@users.noreply.github.com>

godobyte commented on src/deadline_worker_agent/sessions/attachment_models.py:
```diff
@@ -0,0 +1,167 @@
+        manifest_props_data = data["manifestProperties"]
+        from deadline.job_attachments.models import PathFormat
+
+        manifest_properties = ManifestProperties(
+            rootPath=manifest_props_data["rootPath"],
+            rootPathFormat=PathFormat(manifest_props_data["rootPathFormat"]),
+            fileSystemLocationName=manifest_props_data.get("fileSystemLocationName"),
+            inputManifestPath=manifest_props_data.get("inputManifestPath"),
+            inputManifestHash=manifest_props_data.get("inputManifestHash"),
+            outputRelativeDirectories=manifest_props_data.get("outputRelativeDirectories"),
+        )
```
As per agreed, I'll add this `from_dict` to the `ManifestProperties` class and referencing it from `WorkerManifestProperties` in follow-up PRs.

jusiskin commented on test/unit/sessions/test_attachment_models.py:
```diff
@@ -0,0 +1,518 @@
+        )
+
+        with pytest.raises(ValueError, match="local_root_path must be set"):
+            WorkerManifestProperties(manifest_properties=manifest_props, local_root_path="")
```
nit: this is a duplicate test of lines 40-47

godobyte replied:
Good catch, will update.

jusiskin commented on test/unit/sessions/test_attachment_models.py:
```diff
@@ -0,0 +1,518 @@
+        # Verify each instance has its own empty list
+        assert worker_props1.local_manifest_paths == []
+        assert worker_props2.local_manifest_paths == []
+        assert worker_props1.local_manifest_paths is not worker_props2.local_manifest_paths
```
This is a good test to have. It prevents a common Python pitfall where the default is a reference to a single list

jusiskin commented on test/unit/sessions/test_attachment_models.py:
```diff
@@ -0,0 +1,518 @@
+
+        # Verify the provided list is used
+        assert worker_props.local_manifest_paths == provided_paths
+        # Verify it's the same reference (not a copy)
+        assert worker_props.local_manifest_paths is provided_paths
```
Is this the behaviour we want?

This means that both `WorkerManifestProperties` and the caller can modify the list and affect one another. Usually copy semantics are preferred so that `WorkerManifestProperties` maintains its own separately controlled list and is unaffected by changes made by the caller to the list supplied when creating the instance.

godobyte replied:
Absolutely! I was also thinking about that and decided it could be updated later along with the model usage code. Since you also brought it up, I've updated the implementation to use copy semantics for better encapsulation.

jusiskin commented on test/unit/sessions/test_attachment_models.py:
```diff
@@ -0,0 +1,518 @@
+        assert "manifest" not in result_dict
+
+        # Verify manifestProperties content
+        manifest_data = result_dict["manifestProperties"]
+        assert manifest_data["rootPath"] == "/source/path"
+        assert manifest_data["rootPathFormat"] == "posix"
+        assert manifest_data["fileSystemLocationName"] == "shared_storage"
+        assert manifest_data["inputManifestPath"] == "input.json"
+        assert manifest_data["inputManifestHash"] == "hash123"
+        assert manifest_data["outputRelativeDirectories"] == ["out1", "out2"]
```
nit: we are now testing the behavior of `ManifestProperties.to_dict()` here.

Do we want to guard against changes to the `ManifestProperties` from `deadline-cloud`? Consider the case where a new field is added to it. The test would currently pass, but not have coverage for the new fields. Is that a problem?

Is it better to use a mock here if we defer to the behaviour of `ManifestProperties.to_dict(...)`?

godobyte replied:
Great point! The current test is tightly coupled to the implementation details of `ManifestProperties.to_dict()` from the deadline-cloud. I don't think change in deadline-cloud would easily break this since any change to the function is supposed to be backward compatible, but this test does expose implementation detail from the function. I'll update to use mock.

jusiskin replied:
I still see coverage of the `ManifestProperties.to_dict()` in the tests and do not see any mocks added.

jusiskin commented on test/unit/sessions/test_attachment_models.py:
nit: the tests in this file do not follow the `GIVEN` / `WHEN` / `THEN` convention of other test modules

jusiskin replied:
It appears there are still some without this, but I don't consider these merge-blocking. I've pointed out the ones remaining as new review comments, but they are optional.

jusiskin reviewed (changes requested):
Great job with the code here and comprehensive tests. Just a few minor suggestions for improvement.

dependabot[bot] committed:
chore(deps): update ruff requirement from ~=0.12.9 to ~=0.13.0 (#734)

Updates the requirements on [ruff](https://github.com/astral-sh/ruff) to permit the latest version.
- [Release notes](https://github.com/astral-sh/ruff/releases)
- [Changelog](https://github.com/astral-sh/ruff/blob/main/CHANGELOG.md)
- [Commits](https://github.com/astral-sh/ruff/compare/0.12.9...0.13.0)

---
updated-dependencies:
- dependency-name: ruff
  dependency-version: 0.13.0
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Signed-off-by: Godot Bian <13778003+godobyte@users.noreply.github.com>

dependabot[bot] committed:
chore(deps): update twine requirement from ~=6.1 to ~=6.2 (#730)

Updates the requirements on [twine](https://github.com/pypa/twine) to permit the latest version.
- [Release notes](https://github.com/pypa/twine/releases)
- [Changelog](https://github.com/pypa/twine/blob/main/docs/changelog.rst)
- [Commits](https://github.com/pypa/twine/compare/6.1.0...6.2.0)

---
updated-dependencies:
- dependency-name: twine
  dependency-version: 6.2.0
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Signed-off-by: Godot Bian <13778003+godobyte@users.noreply.github.com>

dependabot[bot] committed:
chore(deps): update pytest-cov requirement from ==6.2.* to ==7.0.* (#732)

Updates the requirements on [pytest-cov](https://github.com/pytest-dev/pytest-cov) to permit the latest version.
- [Changelog](https://github.com/pytest-dev/pytest-cov/blob/master/CHANGELOG.rst)
- [Commits](https://github.com/pytest-dev/pytest-cov/compare/v6.2.0...v7.0.0)

---
updated-dependencies:
- dependency-name: pytest-cov
  dependency-version: 7.0.0
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Signed-off-by: Godot Bian <13778003+godobyte@users.noreply.github.com>

dependabot[bot] committed:
chore(deps): update python-semantic-release requirement (#733)

Updates the requirements on [python-semantic-release](https://github.com/python-semantic-release/python-semantic-release) to permit the latest version.
- [Release notes](https://github.com/python-semantic-release/python-semantic-release/releases)
- [Changelog](https://github.com/python-semantic-release/python-semantic-release/blob/master/CHANGELOG.rst)
- [Commits](https://github.com/python-semantic-release/python-semantic-release/compare/v10.3...v10.4)

---
updated-dependencies:
- dependency-name: python-semantic-release
  dependency-version: 10.4.0
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Signed-off-by: Godot Bian <13778003+godobyte@users.noreply.github.com>

godobyte committed:
address comments

Signed-off-by: Godot Bian <13778003+godobyte@users.noreply.github.com>

godobyte committed:
fix mypy error

Signed-off-by: Godot Bian <13778003+godobyte@users.noreply.github.com>

godobyte commented on src/deadline_worker_agent/sessions/attachment_models.py:
```diff
@@ -0,0 +1,173 @@
+    information about local file system paths for manifest files and root directories.
+    """
+
+    def __init__(
+        self,
+        manifest_properties: ManifestProperties,
+        local_root_path: str,
+        local_manifest_paths: Optional[List[str]] = None,
+    ):
+        """
+        Initialize WorkerManifestProperties.
+        Args:
+            manifest_properties: The original manifest properties
+            local_root_path: Local root path for attachment files
+            local_manifest_paths: Optional list of local paths for manifest files (supports step dependencies)
+        """
+        self.manifest_properties = manifest_properties
+        self.local_root_path = local_root_path
+        self.local_manifest_paths = list(local_manifest_paths) if local_manifest_paths else []
```
This seems to be causing mypy error that I cannot produce locally 
```
(deadline-cloud-worker-agent) yuanbian@bcd07416ac9b ~/L/C/W/c/t/b/deadline-cloud-worker-agent (sync-data-model)> mypy src test
Success: no issues found in 176 source files
```

```
cmd [3] | mypy src test
/Users/runner/Library/Application Support/hatch/env/virtual/deadline-cloud-worker-agent/O0kGFt_B/deadline-cloud-worker-agent/lib/python3.11/site-packages/pydantic/v1/env_settings.py:23: error: INTERNAL ERROR -- Please try using mypy master on GitHub:
https://mypy.readthedocs.io/en/stable/common_issues.html#using-a-development-mypy-build
If this issue continues with mypy master, please report a bug at https://github.com/python/mypy/issues
version: 1.18.1
/Users/runner/Library/Application Support/hatch/env/virtual/deadline-cloud-worker-agent/O0kGFt_B/deadline-cloud-worker-agent/lib/python3.11/site-packages/pydantic/v1/env_settings.py:23: : note: please use --show-traceback to print a traceback when reporting a bug
Error: Process completed with exit code 2.
```

I did some research, apparently it's related to mypy and pydantic v1 and v2. I may timebox it and revert to previous implementation if it doesn't work.

godobyte replied:
Turned out to be a breaking change introduced by mypy v1.18.1, fixing in a separate PR https://github.com/aws-deadline/deadline-cloud-worker-agent/pull/736

godobyte committed:
pydantic mypy fix test

Signed-off-by: Godot Bian <13778003+godobyte@users.noreply.github.com>

godobyte committed:
chore(deps): update mypy from ~=1.17 to >= 1.17, <1.18 to avoid breaking change (#736)

Signed-off-by: Godot Bian <13778003+godobyte@users.noreply.github.com>
Co-authored-by: Godot Bian <13778003+godobyte@users.noreply.github.com>
Signed-off-by: Godot Bian <13778003+godobyte@users.noreply.github.com>

godobyte committed:
Revert "pydantic mypy fix test"

This reverts commit 461bf7da3f0e8434eb59f5f8e2a4da267cff840e.

Signed-off-by: Godot Bian <13778003+godobyte@users.noreply.github.com>

godobyte committed:
Merge branch 'mainline' into sync-data-model

jusiskin commented on test/unit/sessions/test_attachment_models.py:
```diff
@@ -0,0 +1,513 @@
+        # Test insertion
+        worker_props.local_manifest_paths.insert(0, "/manifest0.json")
+        assert worker_props.local_manifest_paths[0] == "/manifest0.json"
+        assert len(worker_props.local_manifest_paths) == 4
```
nit: this one is still missing the `GIVEN` / `WHEN` / `THEN`.

Looks like a few more tests below too.

jusiskin commented on test/unit/sessions/test_attachment_models.py:
```diff
@@ -0,0 +1,513 @@
+        assert "outputRelativeDirectories" not in manifest_data
+
+        assert result_dict["localManifestPaths"] == []
+        assert result_dict["localRootPath"] == "/local/root"
```
nit: Same here. This one is still missing the `GIVEN` / `WHEN` / `THEN`

jusiskin commented on test/unit/sessions/test_attachment_models.py:
```diff
@@ -0,0 +1,513 @@
+            "localRootPath": "/local/root",
+        }
+        with pytest.raises(KeyError):
+            WorkerManifestProperties.from_dict(data_missing_manifest)
```
a little unusual in the way pytest does this, but how about this:


```suggestion
        # GIVEN - Missing manifestProperties
        data_missing_manifest = {
            "localRootPath": "/local/root",
        }
        # THEN
        with pytest.raises(KeyError):
            # WHEN
            WorkerManifestProperties.from_dict(data_missing_manifest)
```

same applies to the tests below

jusiskin commented on test/unit/sessions/test_attachment_models.py:
```diff
@@ -0,0 +1,513 @@
+        reconstructed_worker_props = WorkerManifestProperties.from_dict(data_dict)
+
+        # Verify they are equal
+        assert reconstructed_worker_props == original_worker_props
```
nit: missing the `GIVEN` / `WHEN` / `THEN`

jusiskin commented on test/unit/sessions/test_attachment_models.py:
```diff
@@ -0,0 +1,513 @@
+        }
+
+        with pytest.raises(ValueError):
+            WorkerManifestProperties.from_dict(data)
```
nit: missing the `GIVEN` / `WHEN` / `THEN`

jusiskin reviewed (changes requested):
Thanks for addressing prior feedback.

The remaining comments are mostly just nitpicks about the `GIVEN` / `WHEN` / `THEN` comments in the unit tests which I don't consider critical.

The one open comment that I think we want to address is about the remaining test coverage of `ManifestProperties.to_dict()` and prior discussion to use a mock instead.

godobyte committed:
address comments

Signed-off-by: Godot Bian <13778003+godobyte@users.noreply.github.com>

godobyte committed:
Merge branch 'mainline' into sync-data-model
